aten/src/ATen/BatchingUtils.h

```diff
@@ -0,0 +1,39 @@
+#include <ATen/BatchedTensorImpl.h>
+
+namespace at {
+
+/*
+ * Utility functions used to implement batching rules.
+ *
+ * NB: All of these do NOT accept Tensors backed by BatchedTensor, unless
+ * otherwise specified. These APIs usually operate on "unpacked BatchedTensors",
+ * i.e. a (value Tensor, BatchDims) pair. This is for performance reasons:
+ * we do not want to always wrap and unwrap BatchedTensors; we try to
+ * only unwrap once per input tensor per operator and wrap once per output
+ * tensor per operator.
+ */ 
+
+// If the input is a Tensor backed with a BatchedTensorImpl, then
+// this function returns the underlying Tensor and BatchDims.
+// If the input is a Tensor backed with regular TensorImpl, then
+// this function returns the tensor and empty BatchDims.
+TORCH_API std::pair<Tensor, BatchDimsRef> unpackBatched(const Tensor& self);
```

Comment:
It's weird that this returns an owning reference to the tensor, but a reference to batch dims. If the validity of the output argument is going to rely on the input tensor staying live, might as well just return references for both?